### PR TITLE
Indirect - SofQW Moments Scan

### LINKED ISSUE #17986: Indirect - SofQWMomentScan
MSG have asked for some scripts that speed up and to a certain extent automate workflow for data reduction from IRIS and OSIRIS. One of these is a script that reduces raw data, runs it through SofQW and then SofQWMoments.

Also closes #12270

### PR DESCRIPTION
**Pull request by louisemccann**
MSG have asked for an algorithm that reduces .raw files and then runs though [SofQW](http://docs.mantidproject.org/nightly/algorithms/SofQW-v1.html) and then [SofQWMoments](http://docs.mantidproject.org/nightly/algorithms/SofQWMoments-v1.html), in order to create a 'quick run' of the SofQW workflow. This algorithm also produces width and diffusion workspaces. 

**To test:**

Input data: [OSIRIS00100320.zip](https://github.com/mantidproject/mantid/files/702270/OSIRIS00100320.zip)
Ensure Facility=`ISIS`, Instrument = `OSIRIS`

**Part 1:**
```
ISISIndirectEnergyTransfer(InputFiles='OSIRIS00100320.nxs', Instrument='OSIRIS', Analyser='graphite', Reflection='002', Efixed=1.845, SpectraRange='963,1004', FoldMultipleFrames=False, OutputWorkspace='IndirectEnergyTransfer_Workspaces')

Rebin(InputWorkspace='osiris100320_graphite002_red', OutputWorkspace='osiris100320_graphite002_r', Params='-0.4,0.00001,0.4')

SofQW(InputWorkspace='osiris100320_graphite002_r', OutputWorkspace='osiris100320_graphite002_sqw', QAxisBinning='0,0.1,2', EMode='Indirect', EFixed=1.845, Method='Polygon')

SofQWMoments(InputWorkspace='osiris100320_graphite002_sqw', EnergyMin=-0.4, EnergyMax=0.4, OutputWorkspace='osiris100320_graphite002_Moments')
```

**Part 2:**
`SofQWMomentsScan(InputFiles='OSIRIS100320', Instrument='OSIRIS', Analyser='graphite', Reflection='002', SpectraRange='963,1004', QRange='0,0.1,2', EnergyRange='-0.4,0.01,0.4')`

Check that `osiris100320_graphite002_sqw` and `osiris100320_graphite002_moments` are the same in part 2 as produced in part 1

Fixes #17986.

[Release Notes](https://github.com/mantidproject/mantid/blob/559e2b052b4e40de9ff0af9f5897162e1a165305/docs/source/release/v3.9.0/indirect_inelastic.rst)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SofQWMomentsScan.py [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+            if temp is not None:
+                temperatures.append(temp)
+            else:
+                # Get the run number
+                run_no = self._get_InstrRun(input_ws)[1]
+                run_numbers.append(run_no)
+
+        y_axis = mtd[input_workspace_names[0] + '_sqw'].getAxis(1)
+        y_values = y_axis.extractValues()
+        q = list()
```
Comment (AnthonyLim23) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SofQWMomentsScan.py]: q is not a descriptive name (the only thing I can think of is charge, in which case a better variable name would be charge).

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SofQWMomentsScan.py [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+        x_axis_is_temp = len(input_workspace_names) == numb_temp
+
+        if x_axis_is_temp:
+            logger.information('X axis is in temperature')
+            unit = ('Temperature', 'K')
+        else:
+            logger.information('X axis is in run number')
+            unit = ('Run No', 'last 3 digits')
+
+        xdat = list()
```
Comment (AnthonyLim23) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SofQWMomentsScan.py]: xdata etc.

**Comment by SpencerHowells** [comment]
Q is Momentum Transfer a fundamental parameter in neutron scattering. See the Unit Factory!

**Comment by AnthonyLim23** [comment]
If someone else could review this today that would be great.

**Comment by raquelalvarezbanos** [comment]
Workspaces `osiris100320_graphite002_red_Moments` (from part 1) and `osiris100320_graphite002_red_sqw_mom` (from part 2) match.

Workspaces `osiris100320_graphite002_sqw` (from part 1) and `osiris100320_graphite002_red_sqw` (from part 2) match if I change the `Rebin` step in part 1 to:

```
Rebin(InputWorkspace='osiris100320_graphite002_red', 
OutputWorkspace='osiris100320_graphite002_r', Params='-0.4,0.01,0.4')
```

(I assume this is what you meant).

**Comment by raquelalvarezbanos** [comment]
Looks good (code, documentation and results). :shipit:

**Pull request merged**